FLT/DedekindDomain/FiniteAdeleRing/BaseChange.lean

```diff
@@ -491,42 +495,176 @@ noncomputable def tensorAdicCompletionIntegersTo (v : HeightOneSpectrum A) :
+  refine ⟨?_, ?_, ?_⟩
+  . intro t ⟨g, hg, ht⟩
+    -- We have `t = equiv g = ∑ i, b i ⊗ g i`, since `g in ∏ 𝒪_v` and
+    -- `b i ∈ (algebraMap B L).range`, this is `tensorAdicCompletionIntegersTo`
+    -- of some element of `B ⊗[A] 𝒪_v`
+    have hf : ∀ (i : ι), ∃ (w : B), (algebraMap B L w) = (b i) := by
+      intro i
+      apply IsIntegralClosure.isIntegral_iff.mp (hb i)
+    obtain ⟨f, hf_prop⟩ := Classical.axiomOfChoice hf
+    have hf_prop' : ∀ (i : ι), (algebraMap B (L ⊗[K] adicCompletion K v) (f i)) = (b i) ⊗ₜ 1 := by
+      intro i
+      rw [Algebra.TensorProduct.algebraMap_apply, hf_prop]
+    use ∑ (i : ι), (f i) ⊗ₜ ⟨g i, hg i trivial⟩
+    let _ : NonAssocSemiring (B ⊗[A] (adicCompletionIntegers K v)) :=
+      Algebra.TensorProduct.instNonAssocSemiring
+    let _ : AddMonoidHomClass (B ⊗[A] (adicCompletionIntegers K v) →+* L ⊗[K] adicCompletion K v)
+        (B ⊗[A] (adicCompletionIntegers K v)) (L ⊗[K] adicCompletion K v) :=
+      RingHomClass.toAddMonoidHomClass
+    rw [map_sum, ← ht]
+    unfold equiv
+    rw [ContinuousLinearEquiv.restrictScalars_symm_apply,
+      IsModuleTopology.continuousLinearEquiv_symm_apply,
+      Basis.equivFun_symm_apply]
+    apply Finset.sum_congr rfl
+    intro x
+    have : (algebraMap _ (L ⊗[K] adicCompletion K v)) (g x) = 1 ⊗ₜ[K] (g x) := rfl
+    simp [Algebra.smul_def, Algebra.ofId_apply, tensorAdicCompletionIntegersTo, hf_prop',
+        b', this]
+  . rw [ContinuousLinearEquiv.image_symm_eq_preimage]
+    apply IsOpen.preimage equiv.continuous
+    apply isOpen_set_pi Set.finite_univ
+    rintro i -
+    exact Valued.valuationSubring_isOpen (v.adicCompletion K)
+  simp [ValuationSubring.zero_mem]
+
+omit [Algebra.IsSeparable K L] [IsDomain B] [Algebra.IsIntegral A B] [Module.Finite A B]
+    [IsDedekindDomain B] [IsFractionRing B L] in
+lemma tensorAdicCompletionIntegersToRange_eq_closureIntegers :
+    SetLike.coe (tensorAdicCompletionIntegersTo A K L B v).range =
+        closure (algebraMap B (L ⊗[K] adicCompletion K v)).range := by
+  apply Set.Subset.antisymm
+  . apply tensorAdicCompletionIntegersToRange_subset_closureIntegers
+  apply closure_minimal
+  . intro c ⟨b, hb⟩
+    rw [← hb]
+    apply algebraMap_mem
+  apply tensorAdicCompletionIsClosedRange
+
+omit [Algebra A L] [IsScalarTower A B L] [IsIntegralClosure B A L] [Module.Finite A B] in
+lemma prodAdicCompletionsIntegers_eq_closureIntegers :
+    SetLike.coe (Subalgebra.pi (R := B) Set.univ
+      (fun (w : Extension B v) ↦ adicCompletionIntegersSubalgebra L w.1)) =
+    closure (SetLike.coe (algebraMap B _).range) := by
+  rw [Subalgebra.coe_pi]
+  show (SetLike.coe (Submodule.pi _  _)) = _
+  rw [Submodule.coe_pi]
+  let val := (fun (w : Extension B v) ↦ w.1)
+  have hinj : Function.Injective val :=
+    (Set.injective_codRestrict Subtype.property).mp fun _ _ a ↦ a
+  rw [closureAlgebraMapIntegers_eq_prodIntegers L val hinj]
+  rfl
+
+set_option synthInstance.maxHeartbeats 40000 in
```

Comment:
Do you know why this is happening? I can take a look later.

Reply:
Adding
```lean
instance : MulActionHomClass
    (L ⊗[K] adicCompletion K v →ₐ[L] (w : Extension B v) → adicCompletion L w.1) B
    (L ⊗[K] adicCompletion K v) ((w : Extension B v) → adicCompletion L w.1) where
  map_smulₛₗ φ b x := by
    rw [← IsScalarTower.algebraMap_smul L, AlgHom.map_smul_of_tower,
      IsScalarTower.algebraMap_smul, id_def]
```
before the lemma fixes it. I diagnosed this by looking at the instance trace; Lean was spending 1.5 seconds trying and failing to solve this using typeclass inference.